// @boost/log/lib/constants.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var chalk = require('chalk');

var translate = require('./translate.js');

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    'default': e
  };
}

var chalk__default = /*#__PURE__*/_interopDefault(chalk); // In order of priority!


const LOG_LEVELS = ['log', 'trace', 'debug', 'info', 'warn', 'error'];
const DEFAULT_LABELS = {
  debug: chalk__default['default'].gray(translate.msg('log:levelDebug')),
  error: chalk__default['default'].red(translate.msg('log:levelError')),
  info: chalk__default['default'].cyan(translate.msg('log:levelInfo')),
  log: chalk__default['default'].yellow(translate.msg('log:levelLog')),
  trace: chalk__default['default'].magenta(translate.msg('log:levelTrace')),
  warn: chalk__default['default'].yellow(translate.msg('log:levelWarn'))
};
const MAX_LOG_SIZE = 10485760;
exports.DEFAULT_LABELS = DEFAULT_LABELS;
exports.LOG_LEVELS = LOG_LEVELS;
exports.MAX_LOG_SIZE = MAX_LOG_SIZE;
