// @boost/log/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var constants = require('./constants.js');

var createLogger = require('./createLogger.js');

var formats$1 = require('./formats.js');

var Logger = require('./Logger.js');

var Transport = require('./Transport.js');

var ConsoleTransport = require('./transports/ConsoleTransport.js');

var FileTransport = require('./transports/FileTransport.js');

var RotatingFileTransport = require('./transports/RotatingFileTransport.js');

var StreamTransport = require('./transports/StreamTransport.js');

require('./types.js');
/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */


exports.DEFAULT_LABELS = constants.DEFAULT_LABELS;
exports.LOG_LEVELS = constants.LOG_LEVELS;
exports.MAX_LOG_SIZE = constants.MAX_LOG_SIZE;
exports.createLogger = createLogger.createLogger;
exports.formats = formats$1;
exports.Logger = Logger.Logger;
exports.Transport = Transport.Transport;
exports.ConsoleTransport = ConsoleTransport.ConsoleTransport;
exports.FileTransport = FileTransport.FileTransport;
exports.RotatingFileTransport = RotatingFileTransport.RotatingFileTransport;
exports.StreamTransport = StreamTransport.StreamTransport;
