// @boost/config/lib/Configuration.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var common = require('@boost/common');

var event = require('@boost/event');

var Cache = require('./Cache.js');

var ConfigFinder = require('./ConfigFinder.js');

var IgnoreFinder = require('./IgnoreFinder.js');

var Processor = require('./Processor.js');

class Configuration extends common.Contract {
  /**
   * Called after config files are loaded but before processed. Can modify config file list.
   * @category Events
   */

  /**
   * Called after ignore files are loaded. Can modify ignore file list.
   * @category Events
   */

  /**
   * Called after config files are loaded and processed.
   * @category Events
   */
  constructor(name, resolver) {
    super();
    this.onLoadedConfig = new event.WaterfallEvent('loaded-config');
    this.onLoadedIgnore = new event.WaterfallEvent('loaded-ignore');
    this.onProcessedConfig = new event.Event('processed-config');
    this.cache = void 0;
    this.configFinder = void 0;
    this.ignoreFinder = void 0;
    this.processor = void 0;
    this.cache = new Cache.Cache();
    this.configFinder = new ConfigFinder.ConfigFinder({
      name,
      resolver
    }, this.cache);
    this.ignoreFinder = new IgnoreFinder.IgnoreFinder({
      name
    }, this.cache);
    this.processor = new Processor.Processor({
      name
    });
    this.bootstrap();
  }
  /**
   * Clear all cache.
   */


  clearCache() {
    this.clearFileCache();
    this.clearFinderCache();
    return this;
  }
  /**
   * Clear all cached file contents.
   */


  clearFileCache() {
    this.cache.clearFileCache();
    return this;
  }
  /**
   * Clear all cached directory and file path information.
   */


  clearFinderCache() {
    this.cache.clearFinderCache();
    return this;
  }
  /**
   * Traverse upwards from the branch directory, until the root directory is found,
   * or we reach to top of the file system. While traversing, find all config files
   * within each branch directory, and the root.
   */


  async loadConfigFromBranchToRoot(dir) {
    const configs = await this.getConfigFinder().loadFromBranchToRoot(dir);
    return this.processConfigs(this.onLoadedConfig.emit(configs));
  }
  /**
   * Load config files from the defined root. Root is determined by a relative
   * `.config` folder and `package.json` file.
   */


  async loadConfigFromRoot(dir = process.cwd()) {
    const configs = await this.getConfigFinder().loadFromRoot(dir);
    return this.processConfigs(this.onLoadedConfig.emit(configs));
  }
  /**
   * Traverse upwards from the branch directory, until the root directory is found,
   * or we reach to top of the file system. While traversing, find all ignore files
   * within each branch directory, and the root.
   */


  async loadIgnoreFromBranchToRoot(dir) {
    const ignores = await this.getIgnoreFinder().loadFromBranchToRoot(dir);
    return this.onLoadedIgnore.emit(ignores);
  }
  /**
   * Load ignore file from the defined root. Root is determined by a relative
   * `.config` folder and `package.json` file.
   */


  async loadIgnoreFromRoot(dir = process.cwd()) {
    const ignores = await this.getIgnoreFinder().loadFromRoot(dir);
    return this.onLoadedIgnore.emit(ignores);
  }
  /**
   * Add a process handler to customize the processing of key-value setting pairs.
   * May only run a processor on settings found in the root of the configuration object.
   * @public
   */


  addProcessHandler(key, handler) {
    this.getProcessor().addHandler(key, handler);
    return this;
  }
  /**
   * Life cycle called on initialization.
   * @public
   */


  bootstrap() {}
  /**
   * Configure the finder instance.
   * @public
   */


  configureFinder(options) {
    this.getConfigFinder().configure(options);
    return this;
  }
  /**
   * Configure the processor instance.
   * @public
   */


  configureProcessor(options) {
    this.getProcessor().configure(options);
    return this;
  }
  /**
   * Return the config file finder instance.
   */


  getConfigFinder() {
    return this.configFinder;
  }
  /**
   * Return the ignore file finder instance.
   */


  getIgnoreFinder() {
    return this.ignoreFinder;
  }
  /**
   * Return the processor instance.
   */


  getProcessor() {
    return this.processor;
  }
  /**
   * Process all loaded config objects into a single config object, and then validate.
   */


  async processConfigs(files) {
    const config = await this.getProcessor().process(this.options, files, this.blueprint(common.predicates));
    this.onProcessedConfig.emit([config]);
    return {
      config,
      files
    };
  }

}

exports.Configuration = Configuration;
