// @boost/config/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var ConfigError = require('./ConfigError.js');

var Configuration = require('./Configuration.js');

var getEnv = require('./helpers/getEnv.js');

var mergeArray = require('./helpers/mergeArray.js');

var mergeExtends = require('./helpers/mergeExtends.js');

var mergeObject = require('./helpers/mergeObject.js');

var mergePlugins = require('./helpers/mergePlugins.js');

var overwrite = require('./helpers/overwrite.js');

var predicates = require('./predicates.js');

require('./types.js');
/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */


exports.ConfigError = ConfigError.ConfigError;
exports.Configuration = Configuration.Configuration;
exports.getEnv = getEnv.getEnv;
exports.mergeArray = mergeArray.mergeArray;
exports.mergeExtends = mergeExtends.mergeExtends;
exports.mergeObject = mergeObject.mergeObject;
exports.mergePlugins = mergePlugins.mergePlugins;
exports.overwrite = overwrite.overwrite;
exports.createExtendsPredicate = predicates.createExtendsPredicate;
exports.createOverridesPredicate = predicates.createOverridesPredicate;
exports.createPluginsPredicate = predicates.createPluginsPredicate;
