// @beemo/cli/lib/setup.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

const debug = require('debug');

const core = require('@beemo/core');

const cli = require('@boost/cli');

const parseSpecialArgv = require('./parseSpecialArgv.js');

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    'default': e
  };
}

const debug__default = /*#__PURE__*/_interopDefault(debug);
/* eslint-disable import/no-extraneous-dependencies, @typescript-eslint/prefer-nullish-coalescing */


const brandName = process.env.BEEMO_BRAND_NAME || 'Beemo';
const binName = process.env.BEEMO_BRAND_BINARY || 'beemo';
const manualURL = process.env.BEEMO_MANUAL_URL || 'https://beemo.dev';
process.env.BOOSTJS_DEBUG_NAMESPACE = binName; // Instrument our CLI for require() performance

if (process.env.TIMING) {
  require('time-require'); // Boost doesn't enable the debugger early enough,
  // so enable it here if we encounter the debug flag.

} else if (process.argv.includes('--debug')) {
  debug__default["default"].enable(`${binName}:*`);
}

const _parseSpecialArgv$par = parseSpecialArgv.parseSpecialArgv(process.argv.slice(2)),
      argv = _parseSpecialArgv$par.argv,
      parallelArgv = _parseSpecialArgv$par.parallelArgv;

const tool = new core.Tool({
  argv,
  projectName: binName
});

const _require = require('@beemo/core/package.json'),
      version = _require.version;

const footer = cli.applyStyle([tool.msg('app:cliEpilogue', {
  manualURL
}), tool.msg('app:poweredBy', {
  version
})].join('\n'), 'muted');
const program = new cli.Program({
  bin: binName,
  footer,
  name: brandName,
  version
});
exports.argv = argv;
exports.parallelArgv = parallelArgv;
exports.program = program;
exports.tool = tool;
