// @beemo/cli/lib/components/RoutineRow.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

const React = require('react');

const ink = require('ink');

const react = require('@boost/cli/react');

const Duration = require('./Duration.js');

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    'default': e
  };
}

const React__default = /*#__PURE__*/_interopDefault(React);
/* eslint-disable no-nested-ternary */


function RoutineRow({
  routine
}) {
  const status = routine.hasPassed() ? 'success' : routine.hasFailed() ? 'failure' : routine.isSkipped() ? 'warning' : 'muted';
  return /*#__PURE__*/React__default["default"].createElement(ink.Box, {
    paddingLeft: routine.depth + (routine.depth > 0 ? 3 : 0),
    paddingRight: 1
  }, routine.depth === 0 && /*#__PURE__*/React__default["default"].createElement(ink.Box, {
    marginRight: 1
  }, /*#__PURE__*/React__default["default"].createElement(react.Style, {
    type: "muted"
  }, `[${routine.index + 1}]`)), /*#__PURE__*/React__default["default"].createElement(ink.Box, null, routine.depth > 0 && /*#__PURE__*/React__default["default"].createElement(react.Style, {
    type: "muted"
  }, "\u2514\u2500 "), /*#__PURE__*/React__default["default"].createElement(react.Style, {
    bold: true,
    type: status
  }, routine.key.toUpperCase())), /*#__PURE__*/React__default["default"].createElement(ink.Box, {
    marginLeft: 1
  }, /*#__PURE__*/React__default["default"].createElement(ink.Text, null, routine.title)), (routine.isComplete() || routine.isRunning()) && /*#__PURE__*/React__default["default"].createElement(ink.Box, {
    marginLeft: 1
  }, /*#__PURE__*/React__default["default"].createElement(Duration.Duration, {
    time: (routine.stopTime || Date.now()) - routine.startTime
  })));
}

exports.RoutineRow = RoutineRow;
