// @beemo/core/lib/constants.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
const STRATEGY_NONE = 'none';
const STRATEGY_COPY = 'copy';
const STRATEGY_CREATE = 'create';
const STRATEGY_REFERENCE = 'reference';
const STRATEGY_NATIVE = 'native';
const STRATEGY_TEMPLATE = 'template';
const STRATEGY_BUFFER = 'buffer';
const STRATEGY_PIPE = 'pipe';
const STRATEGY_STREAM = 'stream';
const KEBAB_PATTERN = /^[a-z]{1}[a-z-]+[a-z]{1}$/u; // Keep in sync with CLI options

const EXECUTE_OPTIONS = {
  '--concurrency': true,
  '--graph': true,
  '--workspaces': true
};
exports.EXECUTE_OPTIONS = EXECUTE_OPTIONS;
exports.KEBAB_PATTERN = KEBAB_PATTERN;
exports.STRATEGY_BUFFER = STRATEGY_BUFFER;
exports.STRATEGY_COPY = STRATEGY_COPY;
exports.STRATEGY_CREATE = STRATEGY_CREATE;
exports.STRATEGY_NATIVE = STRATEGY_NATIVE;
exports.STRATEGY_NONE = STRATEGY_NONE;
exports.STRATEGY_PIPE = STRATEGY_PIPE;
exports.STRATEGY_REFERENCE = STRATEGY_REFERENCE;
exports.STRATEGY_STREAM = STRATEGY_STREAM;
exports.STRATEGY_TEMPLATE = STRATEGY_TEMPLATE;
