// @beemo/core/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

const common = require('@boost/common');

const constants = require('./constants.js');

const ConfigContext = require('./contexts/ConfigContext.js');

const Context = require('./contexts/Context.js');

const DriverContext = require('./contexts/DriverContext.js');

const ScaffoldContext = require('./contexts/ScaffoldContext.js');

const ScriptContext = require('./contexts/ScriptContext.js');

const Driver = require('./Driver.js');

const Script = require('./Script.js');

const Tool = require('./Tool.js');

require('./types.js');
/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */


Object.defineProperty(exports, 'Path', {
  enumerable: true,
  get: function () {
    return common.Path;
  }
});
exports.EXECUTE_OPTIONS = constants.EXECUTE_OPTIONS;
exports.KEBAB_PATTERN = constants.KEBAB_PATTERN;
exports.STRATEGY_BUFFER = constants.STRATEGY_BUFFER;
exports.STRATEGY_COPY = constants.STRATEGY_COPY;
exports.STRATEGY_CREATE = constants.STRATEGY_CREATE;
exports.STRATEGY_NATIVE = constants.STRATEGY_NATIVE;
exports.STRATEGY_NONE = constants.STRATEGY_NONE;
exports.STRATEGY_PIPE = constants.STRATEGY_PIPE;
exports.STRATEGY_REFERENCE = constants.STRATEGY_REFERENCE;
exports.STRATEGY_STREAM = constants.STRATEGY_STREAM;
exports.STRATEGY_TEMPLATE = constants.STRATEGY_TEMPLATE;
exports.ConfigContext = ConfigContext.ConfigContext;
exports.Context = Context.Context;
exports.DriverContext = DriverContext.DriverContext;
exports.ScaffoldContext = ScaffoldContext.ScaffoldContext;
exports.ScriptContext = ScriptContext.ScriptContext;
exports.Driver = Driver.Driver;
exports.Script = Script.Script;
exports.Tool = Tool.Tool;
