// @beemo/cli/lib/commands/CreateConfig.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

const React = require('react');

const cli = require('@boost/cli');

const setup = require('../setup.js');

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    'default': e
  };
}

const React__default = /*#__PURE__*/_interopDefault(React);

var _dec, _class;

let CreateConfig = (_dec = cli.Config('create-config', setup.tool.msg('app:cliCommandConfig'), {
  aliases: ['config'],
  allowVariadicParams: 'names',
  category: 'core'
}), _dec(_class = class CreateConfig extends cli.Command {
  async run(...names) {
    const _await$Promise$resolv = await Promise.resolve().then(function () {
      return require('../components/App.js');
    }),
          App = _await$Promise$resolv.App;

    const pipeline = setup.tool.createConfigurePipeline(this.getArguments(), names); // @ts-expect-error Event type mismatch

    return /*#__PURE__*/React__default["default"].createElement(App, {
      pipeline: pipeline
    });
  }

}) || _class);
exports.CreateConfig = CreateConfig;
