// @beemo/cli/lib/commands/RunDriver.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

const _rollupPluginBabelHelpers = require('../_virtual/_rollupPluginBabelHelpers.js');

const cli = require('@boost/cli');

const setup = require('../setup.js');

const BaseRunCommand = require('./BaseRunCommand.js');

var _dec, _dec2, _class, _class2;

let RunDriver = (_dec = cli.Config('run-driver', setup.tool.msg('app:cliCommandRunDriver'), {
  allowUnknownOptions: true,
  allowVariadicParams: true,
  category: 'core'
}), _dec2 = cli.Arg.Params({
  description: setup.tool.msg('app:cliArgDriverName'),
  label: 'name',
  required: true,
  type: 'string'
}), _dec(_class = (_class2 = class RunDriver extends BaseRunCommand.BaseRunCommand {
  async run(name = '') {
    const pipeline = setup.tool.createRunDriverPipeline(this.getArguments(), name, this.options.parallelArgv);
    return this.renderDriver(pipeline);
  }

  blueprint({
    array,
    string
  }) {
    return {
      parallelArgv: array(array(string()))
    };
  }

}, _rollupPluginBabelHelpers.applyDecoratedDescriptor(_class2.prototype, "run", [_dec2], Object.getOwnPropertyDescriptor(_class2.prototype, "run"), _class2.prototype), _class2)) || _class);
exports.RunDriver = RunDriver;
