// @beemo/cli/lib/commands/RunScript.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

const _rollupPluginBabelHelpers = require('../_virtual/_rollupPluginBabelHelpers.js');

const cli = require('@boost/cli');

const setup = require('../setup.js');

const BaseRunCommand = require('./BaseRunCommand.js');

var _dec, _dec2, _class, _class2;

let RunScript = (_dec = cli.Config('run-script', setup.tool.msg('app:cliCommandRunScript'), {
  aliases: ['run'],
  allowUnknownOptions: true,
  allowVariadicParams: true,
  category: 'core'
}), _dec2 = cli.Arg.Params({
  description: setup.tool.msg('app:cliArgScriptName'),
  label: 'name',
  required: true,
  type: 'string'
}), _dec(_class = (_class2 = class RunScript extends BaseRunCommand.BaseRunCommand {
  async run(name) {
    const pipeline = setup.tool.createRunScriptPipeline(this.getArguments(), name);
    await pipeline.run();
  }

}, _rollupPluginBabelHelpers.applyDecoratedDescriptor(_class2.prototype, "run", [_dec2], Object.getOwnPropertyDescriptor(_class2.prototype, "run"), _class2.prototype), _class2)) || _class);
exports.RunScript = RunScript;
