// @boost/common/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var optimal = require('optimal');

var CommonError = require('./CommonError.js');

var constants = require('./constants.js');

var Contract = require('./Contract.js');

var ExitError = require('./ExitError.js');

require('./helpers/index.js');

var PackageGraph = require('./PackageGraph.js');

var Path = require('./Path.js');

var PathResolver = require('./PathResolver.js');

var Project = require('./Project.js');

var json$1 = require('./serializers/json.js');

var yaml$1 = require('./serializers/yaml.js');

var types = require('./types.js');

var decorators = require('@boost/decorators');

var createBlueprint = require('./helpers/createBlueprint.js');

var deepFreeze = require('./helpers/deepFreeze.js');

var deepMerge = require('./helpers/deepMerge.js');

var formatMs = require('./helpers/formatMs.js');

var instanceOf = require('./helpers/instanceOf.js');

var isEmpty = require('./helpers/isEmpty.js');

var isFilePath = require('./helpers/isFilePath.js');

var isModuleName = require('./helpers/isModuleName.js');

var isObject = require('./helpers/isObject.js');

var isPlainObject = require('./helpers/isPlainObject.js');

var parseFile = require('./helpers/parseFile.js');

var requireModule = require('./helpers/requireModule.js');

var requireTypedModule = require('./helpers/requireTypedModule.js');

var toArray = require('./helpers/toArray.js');

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    'default': e
  };
}

var optimal__default = /*#__PURE__*/_interopDefault(optimal);
/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */


Object.defineProperty(exports, 'optimal', {
  enumerable: true,
  get: function () {
    return optimal__default['default'];
  }
});
Object.defineProperty(exports, 'predicates', {
  enumerable: true,
  get: function () {
    return optimal.predicates;
  }
});
exports.CommonError = CommonError.CommonError;
exports.MODULE_NAME_PART = constants.MODULE_NAME_PART;
exports.MODULE_NAME_PATTERN = constants.MODULE_NAME_PATTERN;
exports.Contract = Contract.Contract;
exports.ExitError = ExitError.ExitError;
exports.PackageGraph = PackageGraph.PackageGraph;
exports.Path = Path.Path;
exports.PathResolver = PathResolver.PathResolver;
exports.Project = Project.Project;
exports.json = json$1;
exports.yaml = yaml$1;
Object.defineProperty(exports, 'LookupType', {
  enumerable: true,
  get: function () {
    return types.LookupType;
  }
});
exports.createBlueprint = createBlueprint.createBlueprint;
exports.deepFreeze = deepFreeze.deepFreeze;
exports.deepMerge = deepMerge.deepMerge;
exports.formatMs = formatMs.formatMs;
exports.instanceOf = instanceOf.instanceOf;
exports.isEmpty = isEmpty.isEmpty;
exports.isFilePath = isFilePath.isFilePath;
exports.isModuleName = isModuleName.isModuleName;
exports.isObject = isObject.isObject;
exports.isPlainObject = isPlainObject.isPlainObject;
exports.parseFile = parseFile.parseFile;
exports.requireModule = requireModule.requireModule;
exports.requireTypedModule = requireTypedModule.requireTypedModule;
exports.toArray = toArray.toArray;
Object.keys(decorators).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return decorators[k];
    }
  });
});
