// @boost/terminal/lib/cursor.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var ansiEscapes = require('ansi-escapes');

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    'default': e
  };
}

var ansiEscapes__default = /*#__PURE__*/_interopDefault(ansiEscapes);
/** Number of columns to move backward. Defaults to `1`. */


const backward = ansiEscapes__default['default'].cursorBackward;
/** Number of rows to move down. Defaults to `1`. */

const down = ansiEscapes__default['default'].cursorDown;
/** Number of columns to move forward. Defaults to `1`. */

const forward = ansiEscapes__default['default'].cursorForward;
/** Hide the cursor. */

const hide = ansiEscapes__default['default'].cursorHide;
/** Move cursor to the next line. */

const nextLine = ansiEscapes__default['default'].cursorNextLine;
/** Return the current cursur position. */

const position = ansiEscapes__default['default'].cursorGetPosition;
/** Move cursor to the previous line. */

const prevLine = ansiEscapes__default['default'].cursorPrevLine;
/** Restores the cursor position/state. */

const restorePosition = ansiEscapes__default['default'].cursorRestorePosition;
/** Saves the cursor position/state. */

const savePosition = ansiEscapes__default['default'].cursorSavePosition;
/** Show the cursor. */

const show = ansiEscapes__default['default'].cursorShow;
/** Move cursor to the start of the current line. */

const startLine = ansiEscapes__default['default'].cursorLeft;
/** Set the relative position of the cursor. */

const to = ansiEscapes__default['default'].cursorMove;
/** Set the absolute position of the cursor, starting from the top left. */

const toAbsolute = ansiEscapes__default['default'].cursorTo;
/** Number of rows to move up. Defaults to `1`. */

const up = ansiEscapes__default['default'].cursorUp;
exports.backward = backward;
exports.down = down;
exports.forward = forward;
exports.hide = hide;
exports.nextLine = nextLine;
exports.position = position;
exports.prevLine = prevLine;
exports.restorePosition = restorePosition;
exports.savePosition = savePosition;
exports.show = show;
exports.startLine = startLine;
exports.to = to;
exports.toAbsolute = toAbsolute;
exports.up = up;
