// @boost/terminal/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var chalk = require('chalk');

var figures = require('figures');

var cursor$1 = require('./cursor.js');

var screen$1 = require('./screen.js');

var text = require('./text.js');

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    'default': e
  };
}

var chalk__default = /*#__PURE__*/_interopDefault(chalk);

var figures__default = /*#__PURE__*/_interopDefault(figures);
/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */


Object.defineProperty(exports, 'style', {
  enumerable: true,
  get: function () {
    return chalk__default['default'];
  }
});
Object.defineProperty(exports, 'figures', {
  enumerable: true,
  get: function () {
    return figures__default['default'];
  }
});
exports.cursor = cursor$1;
exports.screen = screen$1;
exports.ANSI_REGEX = text.ANSI_REGEX;
exports.annotate = text.annotate;
exports.calculateWidth = text.calculateWidth;
exports.hasAnsi = text.hasAnsi;
exports.link = text.link;
exports.sliceAnsi = text.sliceAnsi;
exports.stripAnsi = text.stripAnsi;
exports.truncate = text.truncate;
exports.wrapAnsi = text.wrapAnsi;
