// @boost/terminal/lib/screen.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var ansiEscapes = require('ansi-escapes');

var termSize = require('term-size');

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    'default': e
  };
}

var ansiEscapes__default = /*#__PURE__*/_interopDefault(ansiEscapes);

var termSize__default = /*#__PURE__*/_interopDefault(termSize);
/** Clear the terminal screen. */


const clear = ansiEscapes__default['default'].clearScreen;
/** Clear the whole terminal, including scrollback buffer. */

const clearTerminal = ansiEscapes__default['default'].clearTerminal;
/** Erase the screen and move the cursor the top left position. */

const erase = ansiEscapes__default['default'].eraseScreen;
/** Erase the screen from the current line down to the bottom of the screen. */

const eraseDown = ansiEscapes__default['default'].eraseDown;
/** Erase from the current cursor position to the end of the current line. */

const eraseEndLine = ansiEscapes__default['default'].eraseEndLine;
/** Erase the entire current line. */

const eraseLine = ansiEscapes__default['default'].eraseLine;
/** Erase from the current cursor position up the specified amount of rows. */

const eraseLines = ansiEscapes__default['default'].eraseLines;
/** Erase from the current cursor position to the start of the current line. */

const eraseStartLine = ansiEscapes__default['default'].eraseStartLine;
/** Erase the screen from the current line up to the top of the screen. */

const eraseUp = ansiEscapes__default['default'].eraseUp;
/** Scroll down the screen by 1 line. */

const scrollDown = ansiEscapes__default['default'].scrollDown;
/** Scroll up the screen by 1 line. */

const scrollUp = ansiEscapes__default['default'].scrollUp;
/** Reliably and accurately get the screen size (in `columns` and `rows`). */

const size = termSize__default['default'];
exports.clear = clear;
exports.clearTerminal = clearTerminal;
exports.erase = erase;
exports.eraseDown = eraseDown;
exports.eraseEndLine = eraseEndLine;
exports.eraseLine = eraseLine;
exports.eraseLines = eraseLines;
exports.eraseStartLine = eraseStartLine;
exports.eraseUp = eraseUp;
exports.scrollDown = scrollDown;
exports.scrollUp = scrollUp;
exports.size = size;
