// @boost/cli/lib/components/internal/HelpParams.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var React = require('react');

var ink = require('ink');

var constants = require('../../constants.js');

require('../../helpers/index.js');

var translate = require('../../translate.js');

var Header = require('../Header.js');

var formatType = require('../../helpers/formatType.js');

var getLongestWidth = require('../../helpers/getLongestWidth.js');

var formatDescription = require('../../helpers/formatDescription.js');

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    'default': e
  };
}

var React__default = /*#__PURE__*/_interopDefault(React);

function gatherParamTags(config) {
  const tags = [];

  if (config.required) {
    tags.push(translate.msg('cli:tagRequired'));
  }

  return tags;
}

function HelpParams({
  config,
  params
}) {
  const labels = React.useMemo(() => params.map((p, index) => {
    var _p$label;

    return `${(_p$label = p.label) !== null && _p$label !== void 0 ? _p$label : index} ${formatType.formatType(p)}`;
  }), [params]);
  const labelWidth = getLongestWidth.getLongestWidth(labels);
  const allowVariadic = config === null || config === void 0 ? void 0 : config.allowVariadicParams;
  return /*#__PURE__*/React__default['default'].createElement(React__default['default'].Fragment, null, /*#__PURE__*/React__default['default'].createElement(Header.Header, {
    label: translate.msg('cli:labelParams')
  }), params.map((param, index) => {
    if (param.hidden) {
      return null;
    }

    const desc = param ? formatDescription.formatDescription(param, gatherParamTags(param)) : '';
    return /*#__PURE__*/React__default['default'].createElement(ink.Box, {
      key: `${labels[index]}-${index}`,
      flexDirection: "row",
      paddingLeft: constants.SPACING_COL
    }, /*#__PURE__*/React__default['default'].createElement(ink.Box, {
      flexGrow: 0,
      width: labelWidth + constants.SPACING_COL_WIDE
    }, /*#__PURE__*/React__default['default'].createElement(ink.Text, null, labels[index])), /*#__PURE__*/React__default['default'].createElement(ink.Box, {
      flexGrow: 1
    }, /*#__PURE__*/React__default['default'].createElement(ink.Text, {
      wrap: "wrap"
    }, desc)));
  }), allowVariadic && /*#__PURE__*/React__default['default'].createElement(ink.Box, {
    paddingLeft: constants.SPACING_COL
  }, /*#__PURE__*/React__default['default'].createElement(ink.Text, null, `…${typeof allowVariadic === 'string' ? allowVariadic : ''}`, ' ', formatType.formatType({
    multiple: true,
    type: 'string'
  }))));
}

exports.HelpParams = HelpParams;
