// @boost/cli/lib/components/internal/Wrapper.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var React = require('react');

var ink = require('ink');

var ProgramContext = require('../../ProgramContext.js');

var Failure = require('../Failure.js');

var LogWriter = require('./LogWriter.js');

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    'default': e
  };
}

var React__default = /*#__PURE__*/_interopDefault(React);

class Wrapper extends React__default['default'].Component {
  constructor(...args) {
    super(...args);
    this.state = {
      error: null
    };
    this.value = void 0;
  }

  static getDerivedStateFromError(error) {
    return {
      error
    };
  }

  componentDidCatch(error) {
    this.setState({
      error
    });
  }

  render() {
    const error = this.state.error;
    const _this$props = this.props,
          children = _this$props.children,
          exit = _this$props.exit,
          log = _this$props.log,
          program = _this$props.program,
          errBuffer = _this$props.errBuffer,
          outBuffer = _this$props.outBuffer;

    if (!this.value) {
      this.value = {
        exit,
        log,
        program
      };
    }

    return /*#__PURE__*/React__default['default'].createElement(ProgramContext.ProgramContext.Provider, {
      value: this.value
    }, /*#__PURE__*/React__default['default'].createElement(ink.Box, null, error ? /*#__PURE__*/React__default['default'].createElement(Failure.Failure, {
      binName: program.bin,
      delimiter: program.delimiter,
      error: error
    }) : children, /*#__PURE__*/React__default['default'].createElement(LogWriter.LogWriter, {
      errBuffer: errBuffer,
      outBuffer: outBuffer
    })));
  }

}

exports.Wrapper = Wrapper;
