// packemon/lib/constants.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
const EXTENSIONS = ['.ts', '.tsx', '.js', '.jsx', '.cjs', '.mjs'];
const EXCLUDE = ['node_modules/**', 'tests/**', '__fixtures__/**', '__mocks__/**', '__tests__/**', '*.config.(c|m)?js', '*.(test|spec).(js|ts)x?']; // https://reactnative.dev/docs/javascript-environment
// Based on browserslist: https://github.com/browserslist/browserslist

const NATIVE_TARGETS = {
  legacy: 'iOS 8',
  stable: 'iOS 10',
  current: 'iOS 12',
  experimental: 'iOS 14'
}; // Based on LTS schedule: https://nodejs.org/en/about/releases/

const NODE_SUPPORTED_VERSIONS = {
  legacy: '10.3.0',
  stable: '12.17.0',
  // ESM support not behind a flag
  current: '14.16.0',
  // Includes security fixes
  experimental: '16.0.0'
};
const NPM_SUPPORTED_VERSIONS = {
  legacy: '6.1.0',
  stable: '6.13.0',
  current: ['6.14.0', '7.0.0'],
  experimental: '7.0.0'
}; // Based on browserslist: https://github.com/browserslist/browserslist

const BROWSER_TARGETS = {
  legacy: 'IE 11',
  stable: ['defaults', 'not IE 11'],
  current: ['> 1%', 'not dead'],
  experimental: ['last 2 chrome versions', 'last 2 firefox versions']
};
const STATE_COLORS = {
  pending: 'muted',
  passed: 'success',
  failed: 'failure'
};
const STATE_LABELS = {
  pending: '',
  building: 'Building',
  passed: 'Passed',
  failed: 'Failed'
};
const DEFAULT_FORMATS = {
  browser: ['lib', 'esm'],
  native: ['lib'],
  node: ['mjs']
};
const FORMATS_BROWSER = ['lib', 'esm', 'umd'];
const FORMATS_NATIVE = ['lib'];
const FORMATS_NODE = ['lib', 'mjs', 'cjs'];
const FORMATS = ['lib', 'esm', 'umd', 'mjs', 'cjs'];
const DEFAULT_INPUT = 'src/index.ts';
const DEFAULT_PLATFORM = 'browser';
const PLATFORMS = ['native', 'node', 'browser'];
const DEFAULT_SUPPORT = 'stable';
const SUPPORTS = ['legacy', 'stable', 'current', 'experimental'];
const SUPPORT_PRIORITY = {
  legacy: 0,
  stable: 1,
  current: 2,
  experimental: 3
};
exports.BROWSER_TARGETS = BROWSER_TARGETS;
exports.DEFAULT_FORMATS = DEFAULT_FORMATS;
exports.DEFAULT_INPUT = DEFAULT_INPUT;
exports.DEFAULT_PLATFORM = DEFAULT_PLATFORM;
exports.DEFAULT_SUPPORT = DEFAULT_SUPPORT;
exports.EXCLUDE = EXCLUDE;
exports.EXTENSIONS = EXTENSIONS;
exports.FORMATS = FORMATS;
exports.FORMATS_BROWSER = FORMATS_BROWSER;
exports.FORMATS_NATIVE = FORMATS_NATIVE;
exports.FORMATS_NODE = FORMATS_NODE;
exports.NATIVE_TARGETS = NATIVE_TARGETS;
exports.NODE_SUPPORTED_VERSIONS = NODE_SUPPORTED_VERSIONS;
exports.NPM_SUPPORTED_VERSIONS = NPM_SUPPORTED_VERSIONS;
exports.PLATFORMS = PLATFORMS;
exports.STATE_COLORS = STATE_COLORS;
exports.STATE_LABELS = STATE_LABELS;
exports.SUPPORTS = SUPPORTS;
exports.SUPPORT_PRIORITY = SUPPORT_PRIORITY;
