// packemon/lib/components/PackageRow.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

const React = require('react');

const ink = require('ink');

const react = require('@boost/cli/react');

const ArtifactList = require('./ArtifactList.js');

const Environment = require('./Environment.js');

const useGroupedArtifacts = require('./hooks/useGroupedArtifacts.js');

const _interopDefault = e => e && e.__esModule ? e : {
  default: e
};

const React__default = /*#__PURE__*/_interopDefault(React);
/* eslint-disable react-perf/jsx-no-new-array-as-prop */


function PackageRow({
  package: pkg
}) {
  const _useGroupedArtifacts$ = useGroupedArtifacts.useGroupedArtifacts(pkg),
        envs = _useGroupedArtifacts$.envs,
        groups = _useGroupedArtifacts$.groups,
        ungrouped = _useGroupedArtifacts$.ungrouped;

  return /*#__PURE__*/React__default.default.createElement(ink.Box, {
    flexDirection: "column",
    marginTop: 1
  }, /*#__PURE__*/React__default.default.createElement(ink.Box, {
    flexDirection: "row"
  }, /*#__PURE__*/React__default.default.createElement(ink.Box, null, /*#__PURE__*/React__default.default.createElement(react.Style, {
    bold: true,
    type: "default"
  }, pkg.getName())), envs.length === 1 && /*#__PURE__*/React__default.default.createElement(ink.Box, {
    marginLeft: 1
  }, /*#__PURE__*/React__default.default.createElement(Environment.Environment, {
    type: envs[0]
  }))), /*#__PURE__*/React__default.default.createElement(ArtifactList.ArtifactList, {
    artifacts: ungrouped
  }), Object.entries(groups).map(([env, set]) => /*#__PURE__*/React__default.default.createElement(ArtifactList.ArtifactList, {
    key: env,
    artifacts: [...set],
    environment: env
  })));
}

exports.PackageRow = PackageRow;
