// packemon/lib/components/Build/index.js
'use strict';

const _excluded = ["packemon", "onBuilt"];

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

Object.defineProperty(exports, '__esModule', {
  value: true
});

const React = require('react');

const ink = require('ink');

const react = require('@boost/cli/react');

const useOnMount = require('../hooks/useOnMount.js');

const PackageList = require('../PackageList.js');

const PackageRow = require('../PackageRow.js');

const _interopDefault = e => e && e.__esModule ? e : {
  default: e
};

const React__default = /*#__PURE__*/_interopDefault(React);

function Build(_ref) {
  let packemon = _ref.packemon,
      onBuilt = _ref.onBuilt,
      options = _objectWithoutProperties(_ref, _excluded);

  const _react$useProgram = react.useProgram(),
        exit = _react$useProgram.exit;

  const _React$useState = React.useState([]),
        _React$useState2 = _slicedToArray(_React$useState, 2),
        packages = _React$useState2[0],
        setPackages = _React$useState2[1];

  const _React$useState3 = React.useState([]),
        _React$useState4 = _slicedToArray(_React$useState3, 2),
        staticPackages = _React$useState4[0],
        setStaticPackages = _React$useState4[1];

  const staticNames = React.useRef(new Set()).current;
  const clearLoop = react.useRenderLoop();
  useOnMount.useOnMount(() => {
    let pkgCount = 0; // Save loaded packages

    packemon.onPackagesLoaded.once(pkgs => {
      setPackages(pkgs);
      pkgCount = pkgs.length;
    }); // Add built package to the static list

    const unlistenBuilt = packemon.onPackageBuilt.listen(pkg => {
      if (pkg.isComplete() && !staticNames.has(pkg.getName())) {
        setStaticPackages(pkgs => [...pkgs, pkg]);
        staticNames.add(pkg.getName());
      }

      if (staticNames.size === pkgCount) {
        clearLoop();
      }
    }); // Run the build process on mount

    void packemon.build(options).then(() => {
      unlistenBuilt();
      setStaticPackages(prev => prev.concat(packages.filter(pkg => pkg.isComplete() && !staticNames.has(pkg.getName()))));
      onBuilt === null || onBuilt === void 0 ? void 0 : onBuilt();
    }).catch(exit).finally(clearLoop);
  });
  const runningPackages = packages.filter(pkg => pkg.isRunning());
  return /*#__PURE__*/React__default.default.createElement(React__default.default.Fragment, null, /*#__PURE__*/React__default.default.createElement(ink.Static, {
    items: staticPackages
  }, pkg => /*#__PURE__*/React__default.default.createElement(PackageRow.PackageRow, {
    key: `static-build-${pkg.getName()}`,
    package: pkg
  })), runningPackages.length > 0 && /*#__PURE__*/React__default.default.createElement(ink.Box, {
    flexDirection: "column"
  }, /*#__PURE__*/React__default.default.createElement(react.Header, {
    label: "Building packages for distribution",
    marginBottom: 0
  }), /*#__PURE__*/React__default.default.createElement(PackageList.PackageList, {
    packages: runningPackages
  })));
}

exports.Build = Build;
