// @boost/cli/lib/components/Header.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var React = require('react');

var ink = require('ink');

var constants = require('../constants.js');

var Style = require('./Style.js');

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    'default': e
  };
}

var React__default = /*#__PURE__*/_interopDefault(React);
/**
 * A React component that renders a header/title within a terminal menu.
 */


function Header(props) {
  const label = props.label,
        _props$marginTop = props.marginTop,
        marginTop = _props$marginTop === void 0 ? constants.SPACING_ROW : _props$marginTop,
        _props$marginBottom = props.marginBottom,
        marginBottom = _props$marginBottom === void 0 ? constants.SPACING_ROW : _props$marginBottom,
        _props$type = props.type,
        type = _props$type === void 0 ? 'default' : _props$type;
  return /*#__PURE__*/React__default['default'].createElement(ink.Box, {
    marginBottom: marginBottom,
    marginTop: marginTop
  }, /*#__PURE__*/React__default['default'].createElement(Style.Style, {
    bold: true,
    inverted: true,
    type: type
  }, ` ${label.toLocaleUpperCase()} `));
}

exports.Header = Header;
