// @boost/cli/lib/components/Help.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var React = require('react');

var ink = require('ink');

var constants = require('../constants.js');

require('../helpers/index.js');

var Header = require('./Header.js');

var HelpCommands = require('./internal/HelpCommands.js');

var HelpOptions = require('./internal/HelpOptions.js');

var HelpParams = require('./internal/HelpParams.js');

var HelpUsage = require('./internal/HelpUsage.js');

var formatDescription = require('../helpers/formatDescription.js');

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    'default': e
  };
}

var React__default = /*#__PURE__*/_interopDefault(React); // 	key: K,
// ): (result: Record<K, React.ComponentType<P>>) => { default: React.ComponentType<P> } {
// 	return ({ [key]: component }) => ({ default: component });
// }
// const Commands = React.lazy(() =>
// 	import('./internal/HelpCommands').then(extractDefault('HelpCommands')),
// );
// const Options = React.lazy(() =>
// 	import('./internal/HelpOptions').then(extractDefault('HelpOptions')),
// );
// const Params = React.lazy(() => import('./internal/HelpParams').then(extractDefault('HelpParams')));
// const Usage = React.lazy(() => import('./internal/HelpUsage').then(extractDefault('HelpUsage')));

/**
 * A React component that renders a help menu for a command.
 * Includes all sub-commands, options, params, and more.
 */
// eslint-disable-next-line complexity


function Help(props) {
  const categories = props.categories,
        commands = props.commands,
        delimiter = props.delimiter,
        options = props.options,
        header = props.header,
        params = props.params,
        config = props.config;
  const hasDesc = Boolean(config === null || config === void 0 ? void 0 : config.description);
  const hasUsage = Boolean((config === null || config === void 0 ? void 0 : config.usage) && config.usage.length > 0);
  const hasParams = Boolean(params && params.length > 0);
  const hasCommands = Boolean(commands && Object.keys(commands).length > 0);
  const hasOptions = Boolean(options && Object.keys(options).length > 0);
  return /*#__PURE__*/React__default['default'].createElement(ink.Box, {
    flexDirection: "column"
  }, (hasDesc || hasUsage) && header && /*#__PURE__*/React__default['default'].createElement(Header.Header, {
    label: header
  }), hasDesc && /*#__PURE__*/React__default['default'].createElement(ink.Box, {
    paddingLeft: constants.SPACING_COL
  }, /*#__PURE__*/React__default['default'].createElement(ink.Text, null, formatDescription.formatDescription(config))), hasUsage && /*#__PURE__*/React__default['default'].createElement(HelpUsage.HelpUsage, {
    delimiter: delimiter,
    usage: config === null || config === void 0 ? void 0 : config.usage
  }), hasParams && /*#__PURE__*/React__default['default'].createElement(HelpParams.HelpParams, {
    config: config,
    params: params
  }), hasCommands && /*#__PURE__*/React__default['default'].createElement(HelpCommands.HelpCommands, {
    categories: categories,
    commands: commands
  }), hasOptions && /*#__PURE__*/React__default['default'].createElement(HelpOptions.HelpOptions, {
    categories: categories,
    options: options
  }));
}

exports.Help = Help;
