// @boost/cli/lib/constants.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
/* eslint-disable unicorn/no-unsafe-regex */
// https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string

const VERSION_FORMAT = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/u;
const LOCALE_FORMAT = /^[a-z]{2}(-[A-Z]{2})?$/u;
const EXIT_PASS = 0;
const EXIT_FAIL = 1;
const INTERNAL_OPTIONS = Symbol('options');
const INTERNAL_PARAMS = Symbol('params');
const INTERNAL_PROGRAM = Symbol('program');
const RESERVED_OPTIONS = [// Existing args
'help', 'locale', 'rest', 'unknown', 'version', // Class methods
'exit', 'log', 'register', 'run', // Class properties
'commands', 'commandAliases'];
const SPACING_COL = 1;
const SPACING_COL_WIDE = 2;
const SPACING_ROW = 1;
const DELIMITER = '$ ';
exports.DELIMITER = DELIMITER;
exports.EXIT_FAIL = EXIT_FAIL;
exports.EXIT_PASS = EXIT_PASS;
exports.INTERNAL_OPTIONS = INTERNAL_OPTIONS;
exports.INTERNAL_PARAMS = INTERNAL_PARAMS;
exports.INTERNAL_PROGRAM = INTERNAL_PROGRAM;
exports.LOCALE_FORMAT = LOCALE_FORMAT;
exports.RESERVED_OPTIONS = RESERVED_OPTIONS;
exports.SPACING_COL = SPACING_COL;
exports.SPACING_COL_WIDE = SPACING_COL_WIDE;
exports.SPACING_ROW = SPACING_ROW;
exports.VERSION_FORMAT = VERSION_FORMAT;
