// @boost/args/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var ArgsError = require('./ArgsError.js');

var constants = require('./constants.js');

var format = require('./format.js');

var parse = require('./parse.js');

var ParseError = require('./ParseError.js');

var parseInContext = require('./parseInContext.js');

require('./types.js');

var ValidationError = require('./ValidationError.js');
/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */


exports.ArgsError = ArgsError.ArgsError;
exports.COMMAND_FORMAT = constants.COMMAND_FORMAT;
exports.DEFAULT_BOOLEAN_VALUE = constants.DEFAULT_BOOLEAN_VALUE;
exports.DEFAULT_NUMBER_VALUE = constants.DEFAULT_NUMBER_VALUE;
exports.DEFAULT_STRING_VALUE = constants.DEFAULT_STRING_VALUE;
exports.LONG_OPTION_FORMAT = constants.LONG_OPTION_FORMAT;
exports.OPTION_LIKE = constants.OPTION_LIKE;
exports.SHORT_OPTION_FORMAT = constants.SHORT_OPTION_FORMAT;
exports.SHORT_OPTION_GROUP_FORMAT = constants.SHORT_OPTION_GROUP_FORMAT;
exports.format = format.format;
exports.parse = parse.parse;
exports.ParseError = ParseError.ParseError;
exports.parseInContext = parseInContext.parseInContext;
exports.ValidationError = ValidationError.ValidationError;
