// packemon/lib/commands/Base.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

const _rollupPluginBabelHelpers = require('../_virtual/_rollupPluginBabelHelpers.js');

const cli = require('@boost/cli');

const common = require('@boost/common');

const Packemon = require('../Packemon.js');

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _class, _descriptor, _descriptor2, _descriptor3, _descriptor4, _descriptor5;

let BaseCommand = (_dec = cli.Arg.String('Current working directory to run in', {
  category: 'global'
}), _dec2 = cli.Arg.String('Filter packages to build', {
  category: 'filter'
}), _dec3 = cli.Arg.String('Only generate specific output formats', {
  category: 'filter',
  short: 'f'
}), _dec4 = cli.Arg.String('Only target specific platforms', {
  category: 'filter',
  short: 'p'
}), _dec5 = cli.Arg.Flag('Skip `private` packages', {
  category: 'filter'
}), _dec6 = common.Memoize(), (_class = class BaseCommand extends cli.Command {
  constructor(...args) {
    super(...args);

    _rollupPluginBabelHelpers.initializerDefineProperty(this, "cwd", _descriptor, this);

    _rollupPluginBabelHelpers.initializerDefineProperty(this, "filter", _descriptor2, this);

    _rollupPluginBabelHelpers.initializerDefineProperty(this, "formats", _descriptor3, this);

    _rollupPluginBabelHelpers.initializerDefineProperty(this, "platforms", _descriptor4, this);

    _rollupPluginBabelHelpers.initializerDefineProperty(this, "skipPrivate", _descriptor5, this);
  }

  get packemon() {
    return new Packemon.Packemon(this.cwd || process.cwd());
  }

  getBuildOptions() {
    return {
      filter: this.filter,
      filterFormats: this.formats,
      filterPlatforms: this.platforms,
      skipPrivate: this.skipPrivate
    };
  }

}, (_descriptor = _rollupPluginBabelHelpers.applyDecoratedDescriptor(_class.prototype, "cwd", [_dec], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function () {
    return '';
  }
}), _descriptor2 = _rollupPluginBabelHelpers.applyDecoratedDescriptor(_class.prototype, "filter", [_dec2], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function () {
    return '';
  }
}), _descriptor3 = _rollupPluginBabelHelpers.applyDecoratedDescriptor(_class.prototype, "formats", [_dec3], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function () {
    return '';
  }
}), _descriptor4 = _rollupPluginBabelHelpers.applyDecoratedDescriptor(_class.prototype, "platforms", [_dec4], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function () {
    return '';
  }
}), _descriptor5 = _rollupPluginBabelHelpers.applyDecoratedDescriptor(_class.prototype, "skipPrivate", [_dec5], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function () {
    return false;
  }
}), _rollupPluginBabelHelpers.applyDecoratedDescriptor(_class.prototype, "packemon", [_dec6], Object.getOwnPropertyDescriptor(_class.prototype, "packemon"), _class.prototype)), _class));
exports.BaseCommand = BaseCommand;
