// packemon/lib/commands/Pack.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

const React = require('react');

const cli = require('@boost/cli');

const Build = require('./Build.js');

const _interopDefault = e => e && e.__esModule ? e : {
  default: e
};

const React__default = /*#__PURE__*/_interopDefault(React);

var _dec, _class;

let PackCommand = (_dec = cli.Config('pack', 'Clean, build, and validate packages for distribution'), _dec(_class = class PackCommand extends Build.BuildCommand {
  async run() {
    const _await$Promise$resolv = await Promise.resolve().then(() => require('../components/Pack/index.js')),
          Pack = _await$Promise$resolv.Pack;

    return /*#__PURE__*/React__default.default.createElement(Pack, {
      addEngines: this.addEngines,
      addExports: this.addExports,
      analyze: this.analyze,
      concurrency: this.concurrency,
      declaration: this.declaration,
      declarationConfig: this.declarationConfig,
      filter: this.filter,
      filterFormats: this.formats,
      filterPlatforms: this.platforms,
      packemon: this.packemon,
      skipPrivate: this.skipPrivate,
      timeout: this.timeout
    });
  }

}) || _class);
exports.PackCommand = PackCommand;
